core/src/main/java/org/elasticsearch/ingest/ProcessorFactoryProvider.java

```diff
@@ -17,21 +17,18 @@
  * under the License.
  */
 
-package org.elasticsearch.plugin.ingest;
+package org.elasticsearch.ingest;
 
 import org.elasticsearch.env.Environment;
-import org.elasticsearch.ingest.TemplateService;
-import org.elasticsearch.ingest.processor.Processor;
+
+import java.util.function.BiFunction;
 
 /**
- * The ingest framework (pipeline, processor and processor factory) can't rely on ES specific code. However some
- * processors rely on reading files from the config directory. We can't add Environment as a constructor parameter,
- * so we need some code that provides the physical location of the configuration directory to the processor factories
- * that need this and this is what this processor factory provider does.
+ * Functional interface that allows to create a {@link org.elasticsearch.ingest.Processor.Factory} once all the needed
+ * components are available at a later stage, more specifically the {@link Environment} and the {@link TemplateService}
+ * which some processors need.
  */
 @FunctionalInterface
-interface ProcessorFactoryProvider {
-
-    Processor.Factory get(Environment environment, TemplateService templateService);
+public interface ProcessorFactoryProvider extends BiFunction<Environment, TemplateService, Processor.Factory> {
```

Comment:
I also wonder why the signature of this can't be `Processor apply(Environment, TemplateService, Map<String, Object> config)` so we can safe an indirection?